docs/design/1161-standardize-stdlib.md

```diff
@@ -0,0 +1,247 @@
+new identifiers are added and become harder to understand.
+
+This proposal aims to:
+
+* Establish a convention for how to name identifiers in the standard library.
+* Propose a way to align existing identifiers with the new convention without
+  breaking backwards compatibility.
+
+For the sake of demonstrating the convention, new identifiers are provided as
+examples, but are out of scope of this proposal and should not be considered
+accepted if this proposal is accepted.
+
+## Proposal
+
+The standard library will introduce namespaces for identifiers that group
+similar identifiers together:
+
+* Array functions
+    * `array.concat` (previously `concat`)
+* Conversion functions
+    * `convert.nonsensitive` (previously `nonsensitive`)
+* Decoding functions
+    * `decode.base64` (previously `decode_base64`)
+    * `decode.json` (previously `decode_json`)
+    * `decode.yaml` (previously `decode_yaml`)
+* OS-related functions
+    * `os.env` (previously `env`)
+* String functions
+    * `string.format` (previously `format`)
+    * `string.join` (previously `join`)
+    * `string.replace` (previously `replace`)
+    * `string.split` (previously `split`)
+    * `string.to_lower` (previously `to_lower`)
+    * `string.to_upper` (previously `to_upper`)
+    * `string.trim` (previously `trim`)
+    * `string.trim_prefix` (previously `trim_prefix`)
+    * `string.trim_space` (previously `trim_space`)
+    * `string.trim_suffix` (previously`trim_suffix`)
+
+For identifiers where a namespace has been introduced, the old identifier will
+be marked deprecated for removal for the next major release.
+
+Some identifiers have not been given a namespace because I could not easily
+identify one: `constants`, `coalesce`, and `json_path`. Until a namespace is
+identified (if ever), these identifiers will not be deprecated.
+
+New identifiers **should** be given a namespace; namespacing makes it easier
+for the same identifier to exist in two different namespaces, such as
+`string.join` (for joining multiple strings) alongside a hypothetical
+`path.join` (for joining file system paths).
+
+Introducing a new namespace should be done as a proposal; ideally proposing
+multiple identifiers that would comprise that namespace. Introducing new
+identifiers in an existing namespace should also be done as a proposal, much
+like today.
+
+I've elected for namespaces to use dot separation (`NAMESPACE.IDENTIFIER`) as
+opposed to underscores (`NAMESPACE_IDENTIFIER`) to align with `constants` and
+component names, which both use dot separation for namespacing.
+
+### Example future stdlib
+
+Three of the five namespaces above only have one identifier. However, there are
+many new identifiers we may want to include in the future. Here's an example of
+what these namespaces could eventually look like:
+
+> **NOTE**: These are not part of the proposal and are only provided as an
+> example.
+
+* Array functions
```

Comment:
Just a side-thread. Thoughts on a namespace like `object.` that, for example can allow you to drop fields, map them, filter them, etc. For example `object.drop(my_object, "__address__")` which could be used on maps in lists of targets if we also had `array.map(my_targets, objects.drop(_, "___address___"))` to do some simple relabelling using the `_` as a placeholder for argument ([Scala does it](https://www.learningjournal.guru/article/scala/functions-in-scala/scala-placeholder-syntax/)). It gives you useful functions, without the need to implement a more complex closures syntax. I'd also be open to something like [Python lambda](https://www.w3schools.com/python/python_lambda.asp) syntax. Just in general, if our closures are limited to a single expression, I think it's a good compromise to keep complexity under control and still give power to the users.

Reply:
Yeah, I think that https://github.com/grafana/alloy/issues/156 should consider an option of using single expression syntax, as it may address the use cases while remaining simpler and more constrained. 

That was a deliberate decision made [in Python for lambdas](https://www.artima.com/weblogs/viewpost.jsp?thread=147358) and, similar to Python's philosophy, I would really like to reduce the complexity where we can. I think most of the things we want to do in Alloy could be done with single expressions built on top of a strong standard library.

Reply:
Re: the `_` placeholder - I am not that attached to this idea - it's something that works in Scala, but takes a bit to get used to. Python's lambdas may be better:
```
targets = array.map(my_targets, lambda x: objects.drop(x, "__address__"))
```
of use `func` instead of `lambda`.